### PR TITLE
Add GitHub AppCards example

### PR DESCRIPTION
**Pull request by addisonschultz**
This pull request adds a working example that demonstrates how 2-way sync works between GitHub Issues, Project Cards, and Miro App Cards

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -47,6 +47,12 @@ npx create-miro-app
...
 | [python-flask-starter](examples/rest/python-flask-starter) | This Python/Flask boilerplate will allow to start using the Miro REST API in a few minutes. This sample implements the full Miro authorization (OAuth2 with refresh token) flow.                                                                                                                        |
 
+### Full Stack Apps
+
+|                                             | Description                                                                                                                    |
+| ------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------ |
+| [github-appcards](examples/github-appcards) | This full stack example shows how to build an integration with GitHub that sync data between GitHub Issues and Miro App Cards. |
```
Comment (marcospinello) [inline README.md:54]: Suggestion:

"This full stack example shows how to build an integration with GitHub that syncs data between GitHub issues and Miro app cards."

File: examples/github-appcards/netlify/functions/issues.js [hunk 2]
```diff
@@ -0,0 +1,111 @@
+/* 
+
+  This endpoint updates Miro App Cards when there is an update to their information
```
Comment (marcospinello) [inline examples/github-appcards/netlify/functions/issues.js:3]: Suggestion:

"This endpoint updates title and/or description of Miro app cards when there the corresponding data source is updated in GitHub."

File: examples/github-appcards/netlify/functions/issues.js [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      "id, miroAppCardId::text, gitHubIssueId, miroUserId::text, gitHubUsername, created_at, miroBoardId, gitHubIssueNumber, auth ( access_token )"
+    )
+    .eq("gitHubIssueId", gitHubIssueId);
+
+  //   No Miro App Card Found
+  if (error) {
+    return {
+      statusCode: 200,
+      body: JSON.stringify({
+        message: "No Miro App Card found for this issue",
```
Comment (marcospinello) [inline examples/github-appcards/netlify/functions/issues.js:41]: Suggestion:

"This GitHub issue doesn't have a corresponding Miro app card."

File: examples/github-appcards/netlify/functions/project-cards.js [hunk 4]
```diff
@@ -0,0 +1,168 @@
+/* 
+
+  This endpoint updates Miro App Card statuses when there is an update to their poisition
+  or column in GitHub (i.e. from "In Progress" to "Done")
```
Comment (marcospinello) [inline examples/github-appcards/netlify/functions/project-cards.js:4]: Suggestion:

"This endpoint updates Miro app card statuses when the corresponding GitHub issues change state. For example, from 'In progress' to 'Done'.""

File: examples/github-appcards/netlify/functions/project-cards.js [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+      "id, miroAppCardId::text, gitHubIssueId, miroUserId::text, gitHubUsername, created_at, miroBoardId, gitHubIssueNumber, auth ( access_token )"
+    )
+    .eq("gitHubIssueNumber", gitHubIssueNumer);
+
+  // No Miro App Card Found
+  if (error) {
+    return {
+      statusCode: 200,
+      body: JSON.stringify({
+        message: "No Miro App Card found for this project card",
```
Comment (marcospinello) [inline examples/github-appcards/netlify/functions/project-cards.js:65]: Suggestion:

"This GitHub project card doesn't have a corresponding Miro app card."

File: examples/github-appcards/src/components/GitHub/GitHub.tsx [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+      fullscreen: false,
+    });
+  };
+
+  return (
+    <div id={"github-container"}>
+      <h3>Synced changes</h3>
+      <p>
+        Any changes you make in Miro or GitHub will be synced between both
+        tools.
```
Comment (marcospinello) [inline examples/github-appcards/src/components/GitHub/GitHub.tsx:17]: Suggestion:

"Any changes you apply, either in Miro or in GitHub, are synced between both tools."

File: examples/github-appcards/src/appcard-modal.tsx [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+      });
+  };
+
+  const handleCancelClick = async () => {
+    await miro.board.ui.closeModal();
+  };
+
+  return (
+    <div className="appcard-modal-container">
+      <h1>Edit GitHub Card</h1>
```
Comment (marcospinello) [inline examples/github-appcards/src/appcard-modal.tsx:157]: Can you set "card" to lowercase?

"Edit GitHub card"

File: examples/github-appcards/src/components/Miro/Miro.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+          color={selectedColor}
+        />
+      )}
+
+      <button
+        className="button button-primary"
+        disabled={!itemsSelected}
+        onClick={() => handleCreateGitHubCards(selectedItems)}
+      >
+        Convert to GitHub Card
```
Comment (marcospinello) [inline examples/github-appcards/src/components/Miro/Miro.tsx:95]: Can you set "card" to lowercase?

"Convert to GitHub card"

File: examples/github-appcards/src/components/Miro/NoSelection.tsx [hunk 9]
```diff
@@ -0,0 +1,18 @@
+import * as React from "react";
+import selection from "../../assets/selection.png";
+
+const NoSelection = () => {
+  return (
+    <>
+      <h3>Select from Miro Board</h3>
```
Comment (marcospinello) [inline examples/github-appcards/src/components/Miro/NoSelection.tsx:7]: Can you set "board" to lowercase?

"Select from Miro board"

File: examples/github-appcards/src/components/Modal/GitHubIssueHeader.tsx [hunk 10]
```diff
@@ -0,0 +1,22 @@
...
+    <>
+      <div id={"grid-checkbox"}></div>
+      <div id={"grid-title"}>
+        <p>Issue</p>
+      </div>
+      <div id={"grid-status"}>
+        <p>Status</p>
+      </div>
+      <div id={"grid-date"}>
+        <p>Date Created</p>
```
Comment (marcospinello) [inline examples/github-appcards/src/components/Modal/GitHubIssueHeader.tsx:16]: Can you set "created" to lowercase?

"Date created"

File: examples/github-appcards/src/components/Miro/NoSelection.tsx [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+      <p>
+        Select at least one sticky note or Miro card from the board to convert
+        to a GitHub card.
```
Comment (marcospinello) [inline examples/github-appcards/src/components/Miro/NoSelection.tsx:10]: Suggestion:

"From the board, select at least a sticky note or a card to convert it to a GitHub card."

File: examples/github-appcards/src/assets/style.css [hunk 12]
```diff
@@ -0,0 +1,287 @@
...
+  color: red;
+}
+
+/* 
+
+  Input
+
+*/
+
+.input-container {
```
[the remark this answers is not in the document]
Comment (mettin) [reply examples/github-appcards/src/assets/style.css:199]: If so: create an issue there please

File: examples/github-appcards/src/components/Modal/GitHubIssueHeader.tsx [hunk 13]
```diff
@@ -0,0 +1,22 @@
+import * as React from "react";
+import Checkbox from "../Checkbox";
+import Tag from "../Tag";
+
+const GitHubIssueHeader = () => {
+  return (
+    <>
+      <div id={"grid-checkbox"}></div>
```
Comment (mettin) [inline examples/github-appcards/src/components/Modal/GitHubIssueHeader.tsx:8]: No need for `{` and `}` for all these ids

File: examples/github-appcards/src/components/Modal/GitHubIssueRow.tsx [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+
+const GitHubIssueRow = ({
+  title,
+  date,
+  status,
+  onSelect,
+}: {
+  title: string;
+  date: string;
+  status: { name: string; id: any };
```
Comment (mettin) [inline examples/github-appcards/src/components/Modal/GitHubIssueRow.tsx:14]: can you specify `id`?

File: examples/github-appcards/src/components/Modal/GitHubIssueRow.tsx [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+      const color = await getStatusColor(status.name);
+      setColor(color);
+    }
+
+    generateStatusColor();
+  });
+
+  return (
+    <>
+      <div id={"grid-checkbox"}>
```
Comment (mettin) [inline examples/github-appcards/src/components/Modal/GitHubIssueRow.tsx:35]: No need for `{` and `}`. Use classnames

File: examples/github-appcards/src/components/Select.tsx [hunk 16]
```diff
@@ -0,0 +1,41 @@
...
+const Select = ({
+  label,
+  required,
+  options,
+  selected = { name: "" },
+  onChange,
+}: {
+  label: string;
+  required: boolean;
+  options: any[];
```
Comment (mettin) [inline examples/github-appcards/src/components/Select.tsx:12]:
```suggestion
  options: {name: string}[];
```

File: examples/github-appcards/src/components/Tag.tsx [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+
+const Tag = ({
+  status,
+  color,
+}: {
+  status: { name: string; id: any };
+  color: string;
+}) => {
+  return (
+    <div id={"tag-container"} style={{ backgroundColor: color }}>
```
Comment (mettin) [inline examples/github-appcards/src/components/Tag.tsx:11]: No need for `{` and `}`. Use classnames

File: examples/github-appcards/src/modal.tsx [hunk 18]
```diff
@@ -0,0 +1,201 @@
...
+            ...cards,
+          ]);
+        });
+      });
+    }
+  }, [gitHubColumns]);
+
+  // Fetch GitHub Issues & filter issues that are not in current project
+  React.useEffect(() => {
+    fetchGitHubIssues(username, repo)
```
Comment (mettin) [inline examples/github-appcards/src/modal.tsx:89]: filterGitHubIssues can `throw` but you have no catch

File: examples/github-appcards/src/modal.tsx [hunk 19]
```diff
@@ -0,0 +1,201 @@
...
+
+  // Handle importing and converting GitHub issues to App Cards
+  const handleImportClick = async () => {
+    await insertGitHubAppCards(selectedGitHubIssues).then(() => {
+      miro.board.ui.closeModal();
+    });
+  };
+
+  return (
+    <div id={"modal-container"} className="wrapper">
```
Comment (mettin) [inline examples/github-appcards/src/modal.tsx:158]: No need for `{` and `}`. Use className to style

File: examples/github-appcards/src/modal.tsx [hunk 20]
```diff
@@ -0,0 +1,201 @@
...
+      <h2>Choose from GitHub</h2>
+      <Select
+        label={"Select GitHub Project"}
```
Comment (mettin) [inline examples/github-appcards/src/modal.tsx:161]: No need for `{` and `}`.

File: examples/github-appcards/src/modal.tsx [hunk 21]
```diff
@@ -0,0 +1,201 @@
...
+              return (
+                <GitHubIssueRow
+                  title={issue.title}
+                  date={issue.created_at}
+                  status={issue.status}
+                  onSelect={(value) => handleGitHubIssueSelect(value, issue)}
+                  key={index}
+                />
+              );
+            })}
```
Comment (mettin) [inline examples/github-appcards/src/modal.tsx:185]:
```suggestion
            {filterGitHubIssues().map((issue, index) => (
                <GitHubIssueRow
                  title={issue.title}
                  date={issue.created_at}
                  status={issue.status}
                  onSelect={(value) => handleGitHubIssueSelect(value, issue)}
                  key={index}
                />
              );
            )}
```

File: examples/github-appcards/src/utils/github.ts [hunk 22]
```diff
@@ -0,0 +1,296 @@
+const baseUrl = "https://api.github.com";
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:1]:
```suggestion
const GITHUB_API_URL = "https://api.github.com";
```

File: examples/github-appcards/src/utils/github.ts [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+    `${baseUrl}/repos/${username}/${repo}/projects`,
+    {
+      method: "GET",
+      headers: headers,
+    }
+  )
+    .then((response) => response.json())
+    .then((result) => {
+      const gitHubProjectData = result.map((project: any) => {
+        return {
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:29]: Am i missing something or is this map not doing anything? In any case there's no need to store the results in a const first, you could just `return result.map(....)`

File: examples/github-appcards/src/utils/github.ts [hunk 24]
```diff
@@ -0,0 +1,296 @@
...
+ * Returns {
+ * }
+ */
+export const fetchGitHubColumn = async (column_id: string) => {
+  const gitHubColumn = fetch(`${baseUrl}/projects/columns/${column_id}`, {
+    method: "GET",
+    headers: headers,
+  })
+    .then((response) => response.json())
+    .then((result) => {
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:82]: this step does nothing

File: examples/github-appcards/src/utils/github.ts [hunk 25]
```diff
@@ -0,0 +1,296 @@
...
+  const gitHubProjectCards = fetch(
+    `${baseUrl}/projects/columns/${column_id}/cards`,
+    {
+      method: "GET",
+      headers: headers,
+    }
+  )
+    .then((response) => response.json())
+    .then((result) => {
+      return result;
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:106]: This step does nothing

File: examples/github-appcards/src/utils/github.ts [hunk 26]
```diff
@@ -0,0 +1,296 @@
...
+  const gitHubProjectCard = fetch(
+    `${baseUrl}/projects/columns/cards/${card_id}`,
+    {
+      method: "GET",
+      headers: headers,
+    }
+  )
+    .then((response) => response.json())
+    .then((result) => {
+      return result;
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:129]: This step does nothing

File: examples/github-appcards/src/utils/github.ts [hunk 27]
```diff
@@ -0,0 +1,296 @@
...
+  const gitHubProjectCards = fetch(
+    `${baseUrl}/repos/${username}/${repo}/issues`,
+    {
+      method: "GET",
+      headers: headers,
+    }
+  )
+    .then((response) => response.json())
+    .then((result) => {
+      return result;
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:152]: This step does nothing

File: examples/github-appcards/src/utils/github.ts [hunk 28]
```diff
@@ -0,0 +1,296 @@
...
+  const gitHubCollaborators = fetch(
+    `${baseUrl}/projects/${project_id}/collaborators`,
+    {
+      method: "GET",
+      headers: headers,
+    }
+  )
+    .then((response) => response.json())
+    .then((result) => {
+      return result;
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:175]: This step does nothing

File: examples/github-appcards/src/utils/github.ts [hunk 29]
```diff
@@ -0,0 +1,296 @@
...
+  }
+) => {
+  const gitHubIssue = fetch(`${baseUrl}/repos/${owner}/${repo}/issues`, {
+    method: "POST",
+    headers: headers,
+    body: JSON.stringify(data),
+  })
+    .then((response) => response.json())
+    .then((result) => {
+      return result;
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:203]: This step does nothing

File: examples/github-appcards/src/utils/github.ts [hunk 30]
```diff
@@ -0,0 +1,296 @@
...
+    `${baseUrl}/projects/columns/${column_id}/cards`,
+    {
+      method: "POST",
+      headers: headers,
+      body: JSON.stringify(data),
+    }
+  )
+    .then((response) => response.json())
+    .then((result) => {
+      return result;
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:232]: This step does nothing

File: examples/github-appcards/src/utils/github.ts [hunk 31]
```diff
@@ -0,0 +1,296 @@
...
+    `${baseUrl}/repos/${username}/${repo}/issues/${issueNumber}`,
+    {
+      method: "PATCH",
+      headers: headers,
+      body: JSON.stringify(data),
+    }
+  )
+    .then((response) => response.json())
+    .then((result) => {
+      return result;
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:262]: This step does nothing

File: examples/github-appcards/src/utils/github.ts [hunk 32]
```diff
@@ -0,0 +1,296 @@
...
+    `${baseUrl}/projects/columns/cards/${gitHubProjectCardId}/moves`,
+    {
+      method: "POST",
+      headers: headers,
+      body: JSON.stringify(data),
+    }
+  )
+    .then((response) => response.json())
+    .then((result) => {
+      return result;
```
Comment (mettin) [inline examples/github-appcards/src/utils/github.ts:291]: This step does nothing

File: examples/github-appcards/src/components/Checkbox.tsx [hunk 33]
```diff
@@ -0,0 +1,28 @@
...
+
+  return (
+    <label className="checkbox">
+      <input
+        type="checkbox"
+        tabIndex={0}
+        checked={checked}
+        onChange={handleChange}
+      />
+      <span></span>
```
[the remark this answers is not in the document]
Comment (addisonschultz) [reply examples/github-appcards/src/components/Checkbox.tsx:23]: Getting weird sizing issues with the checkbox without the <span>. It's rendering as 0 x 0

**Comment by addisonschultz** [comment]
Hey @mettin, thanks so much for reviewing 🙏 I've learned a lot about formatting code/examples, and made most of the suggestions/changes you've requested. 

However, I had some questions on finishing up certain files/changes, as I'm not sure the best way/appropriate way to go about the change is. Here are some things I'd like to discuss with you: 

- Properly doing async/await in `netlify/functions/issues.js` (This contains `new Promise` constructors..)
- Properly doing async/await in `netlify/functions/project-cards.js` (This also contains `new Promise` constructors)
- This comment / suggestion https://github.com/miroapp/app-examples/pull/75#discussion_r925672850

Do you have some time this week to hop on a call / discuss?

File: examples/github-appcards/README.md [hunk 34]
```diff
@@ -0,0 +1,72 @@
...
+- Open a board; you should see your app in the app toolbar or in the **Apps** panel.
+
+### How to build the app
+
+- Run `npm run build`. \
+  This generates a static output inside `dist/`, which you can host on a static hosting service.
+
+### Folder structure
+
+The strcuture below outlines the most important files in the project used for it's core functionality.
```
Comment (marcospinello) [inline examples/github-appcards/README.md:30]:
```suggestion
The structure below outlines the most important files in the project used for its core functionality.
```

File: examples/github-appcards/README.md [hunk 35]
```diff
@@ -0,0 +1,72 @@
...
+│      └── style.css <-- CSS styles for the app.
+|  └── components <-- Folder of all UI components used in the app.
+|  └── utils
+│      └── github.ts <-- Utility functions for interacting with the GitHub API
+│          miro.ts <-- Utility functions for interacting with the Miro Web SDK & API
+│          supabase.ts <-- Main entry for interacting with Supabase
+│  └── app.tsx <-- The main entry. Contains structure for the sidebar when launched.
+│      appcard-modal.tsx <-- The main entry for the modal that appears when an app card is expanded.
+│      constants.ts <-- A collection of static variables used throughout the app.
+│      index.ts <-- Initializes app, and contains logic ffor opening the app from the sidebar and expanding an app card.
```
Comment (marcospinello) [inline examples/github-appcards/README.md:47]:
```suggestion
│      index.ts <-- Initializes app, and contains logic for opening the app from the sidebar and expanding an app card.
```

**Review by mettin — APPROVED** [review]

**Pull request merged**